core/arch/aarch64/codec.txt

```diff
@@ -947,8 +947,8 @@ x101101011000000000101xxxxxxxxxx  cls     wx0 : wx5
 
 # Data Processing - Scalar Floating-Point and Advanced SIMD
 # FADD (vector)
-0x001110010xxxxx000101xxxxxxxxxx     fadd vq0 impFPSR : vq5 vq16 impFPCR
-0x0011100x1xxxxx110101xxxxxxxxxx     fadd vq0 impFPSR : vq5 vq16 fsz impFPCR
+0x001110010xxxxx000101xxxxxxxxxx     fadd dq0 : dq5 dq16 fsz16
+0x0011100x1xxxxx110101xxxxxxxxxx     fadd dq0 : dq5 dq16 fsz
 
 # FADD (scalar)
-00011110xx1xxxxx001010xxxxxxxxxx     fadd float_reg0 impFPSR : float_reg5 float_reg16 impFPCR
+00011110xx1xxxxx001010xxxxxxxxxx     fadd float_reg0 : float_reg5 float_reg16
```

Comment:
v8.0 doesn't have half width precision. Bit 22 determines double or single, 23 is 0.

Reply:
Yes, `decode_float_reg` and `encode_float_reg` take care of encoding bit 22 and 23.

Reply:
How do we deal with a client calling a FADD scalar instruction macro with half precision on v8.0 hardware? Is that something that documentation will state is not allowed, or do we check hardware version before executing? I guess this is a particular case of a general versioning issue.

Reply:
I do not think there is much that plain DynamoRIO can do about that, they will be executing an illegal instruction. I am not sure how this is handled on other platforms, but we could provide functions to check if an instruction is available on a given architecture version, so that clients can check before emitting instructions. I think it is reasonable to expect the client to check if the instructions are available, because if it is not available, the client has to insert different instructions (or fail)

Reply:
I don't think anything needs to be done specially: there will be a SIGILL that DynamoRIO will pass on to the app in the same way that it would pass on a SIGILL caused by a sys_kill.

Now, if there were an instruction that has to be mangled into the code cache, some kind of branch instruction, say, that is only available on newer versions of the architecture, then it would be necessary for DynamoRIO to take account of the hardware version, I suppose, to avoid the app seeing an instruction that shouldn't work apparently working.